WIP: add support for single and long double

kohr-h opened the pull request:
This is an attempt to patch in support for single and long double. Upstream supports them since 3.3 or so, and some initial tests with this version look promising. Needs proper testing, though.

Will bad things happen if you run the FFTW thread init functions for all precisions in the same process? Linking all of the libs is no issue.

The goal of this PR is to make a conda forge package for NFFT and pyNFFT, both of which support all the data types.

Closes #64

---

### TODO

- [x] Test with recent NFFT
- [ ] Update documentation
- [ ] Mention API changes in release notes

grlee77 commented:
> Will bad things happen if you run the FFTW thread init functions for all precisions in the same process? 

I don't think so. We call all of them (when available) in pyFFTW: 
https://github.com/pyFFTW/pyFFTW/blob/701350df800b9440b8699d5860276e75effde626/pyfftw/pyfftw.pyx#L653-L658

pyFFTW also recently added support for build and run-time detection for the presence of single, double and/or long-double support. This removed the requirement to have all three variants of FFTW available, but was a good bit of work to implement (see work by @fredRos in pyFFTW/pyFFTW#177 and pyFFTW/pyFFTW#189).

Also, I see that the `setup.py` here (and in master) assumes particular names for the FFTW libraries. If FFTW is built against OpenMP instead of pthreads this assumption will be incorrect, but that could always be addressed in a future PR as needed.

kohr-h commented:
> I don't think so. We call all of them (when available) in pyFFTW:
> https://github.com/pyFFTW/pyFFTW/blob/701350df800b9440b8699d5860276e75effde626/pyfftw/pyfftw.pyx#L653-L658

Ok great!



> pyFFTW also recently added support for build and run-time detection for the presence of single, double and/or long-double support. This removed the requirement to have all three variants of FFTW available, but was a good bit of work to implement (see work by @fredRos in [pyFFTW/pyFFTW#177](https://github.com/pyFFTW/pyFFTW/pull/177) and [pyFFTW/pyFFTW#189](https://github.com/pyFFTW/pyFFTW/pull/189) ).

That sure looks like a lot of work! If the logic can be largely copied, then it should be doable for NFFT as well, but for now I stay away from it :grimacing: 

 

> Also, I see that the `setup.py` here (and in master) assumes particular names for the FFTW libraries. If FFTW is built against OpenMP instead of pthreads this assumption will be incorrect, but that could always be addressed in a future PR as needed.

The fix for that might be a bit easier to do than the one you mentioned above. But yes, not in this PR.

kohr-h commented:
So I've implemented all the basic functionality. The code is a bit repetitive, but I still find it more readable than using name-mangling macros. Maybe when more parts of NFFT come in, it'll be worth refactoring a bit.

Anyway, I've also added tests (rewritten the old ones with `pytest`), and they all pass on my machine (Linux). There's no CI so I need to trust that result.

kohr-h commented:
@grlee77 Are you a maintainer? Or who can pick up this PR?

kohr-h commented:
Oh, before I forget: With single precision, the default `m=12` for the window function cutoff results in `NaN` results. I don't exactly know where that comes from, but with `m=6` it's OK.

grlee77 commented:
> @grlee77 Are you a maintainer? Or who can pick up this PR?

No, I am only a maintainer for the conda-forge feedstock. I have not contributed to pyNFFT itself. I think only @ghisvail can currently merge PRs here.

ghisvail commented:
If any of you is interested in picking up the maintenance of pyNFFT, feel free to contact me. I will create an GitHub organization for it soon, and give commit access to people interested in developing the software further.

kohr-h commented:
I'd like to move forward with this PR. Apart from more extensive tests, I think it's in a decent shape for review. Care to have a look @grlee77?

grlee77 commented:
I can take a look. I also do not currently have merge rights, though.

grlee77 commented on doc/source/api/util.rst:
```diff
@@ -15,10 +15,3 @@ scripts. For instance::
-Functions used for computing the density compensation weights necessary for the
-iterative solver and adjoint NFFT.
-
-.. autofunction:: pynfft.util.voronoi_weights_1d(w, x)
```
see other comment about not removing this 1D function

grlee77 commented on pynfft/cnfft3util.pxd:
```diff
@@ -15,20 +15,19 @@
+    void nfft_vrand_shifted_unit_double(double *x, int n)
         # Inits a vector of random double numbers in $[-1/2,1/2]$ .
 
-    void nfft_voronoi_weights_1d (double *w, double *x, int M)
```
I don't it is necessary to remove `voronoi_weights_1d`. That function still exists in recent releases: https://github.com/NFFT/nfft/blob/cf336307f8794510717584cdbf616bdcde860959/kernel/util/voronoi.c#L23

grlee77 replied:
on second look I don't see that function in `nfft3.h`, but it is in `infft.h`. Those two headers are both included in `api.h`. Note sure if we can just use api.h instead of nfft3.h?

kohr-h replied:
That rings a bell. I think I removed it due to a compiler error I got since the function was nowhere declared in `nfft3.h`.
I'll see if `api.h` is an option, thx for the pointer.

kohr-h replied:
Actually, including `infft.h` as well through `api.h` opens up the possibility to use the name mangling macros, which could help avoiding duplication.

kohr-h replied:
I'll check in the NFFT documentation what is recommended. Perhaps `infft.h` is not intended for use outside of the library, and will cause us a lot of breakage between versions. OTOH, `api.h` sounds pretty public.

grlee77 commented on pynfft/cnfft3util.pxd:
```diff
@@ -15,20 +15,19 @@
+    void nfftf_vrand_unit_complex(fftwf_complex *x, int n)
+    void nfftf_vrand_shifted_unit_double(float *x, int n)
 
-    void nfft_voronoi_weights_S2(double *w, double *xi, int M)
```
removing this should also fix #64 which was closed (I do not currently have the rights to reopen it, but I guess we won't need to if we address it here)

grlee77 commented on pynfft/nfft.pxd:
```diff
@@ -15,27 +15,39 @@
+    object nfft_supported_flags_tuple
+
+
+ctypedef union _nfft_plan:
```
nice. good way to avoid too much duplication

kohr-h replied:
If you know a good way to avoid some of the other duplication as well, please let me know.

kohr-h replied:
> If I understand correctly any given NFFT plan will only have one of these actually initialized via a call to `nfft_init_guru`, etc.

Yes (or rather, the single memory location will be initialized, and only one interpretation will be correct at a given time).

grlee77 commented on pynfft/util.pyx:
```diff
@@ -24,60 +24,47 @@ from cnfft3util cimport *
     '''
-    nfft_vrand_shifted_unit_double(<double *>&x[0], x.size)
-
-def voronoi_weights_1d (object[np.float64_t, mode='c'] w not None,
```
again, potentially keep this one

grlee77 commented:
Also, I didn't look through the docs just now, but we should specify somewhere that all three NFFT variants are now required.

Then when we make a release, remember to also mention that and that support for the 2D Voronoi function has been dropped.

ghisvail commented:
I believe the NFFT utility functions (rand unit and voronoi) should be replaced by native ones using NumPy. Same for the solver, which could be substituted by wrapping a plan in a LinearOperator to be used within SciPy's implementation of the conjugate gradient.

ghisvail commented:
Also @grlee77 and @kohr-h, I have added both of you as collaborators with RW privileges. I must have forgotten back then. Sorry for the lateness, and thank you for taking good care of pynfft.

kohr-h commented:
> I think the only main downside to the implementation is the requirement to have all three variants. That might cause problems for some users (e.g. some compiler/platforms don't support long double?). Given that build/runtime detection is quite a bit of work I am fine to merge this as is and then add that later if needed/requested.

Agreed. If necessary we might want to pick some of the code from pyFFTW, but if I recall correctly it's quite complex.
Regarding availability on Windows, [precompiled binaries](http://fftw.org/install/windows.html) include exactly the precisions we need, and on Mac and Linux, there are Conda packages, distro packages and whatnot. So for now I assume that it's always possible to satisfy the requirements.

> Can you confirm that these are all passing for you? Is this also true for the latest NFFT release?

When I implemented the changes, I tested with the then latest version of NFFT, probably 3.5.0, but not 100 % sure. I'll make sure to test a few versions.



> Also, I didn't look through the docs just now, but we should specify somewhere that all three NFFT variants are now required.

Yes, I didn't take any steps there yet. Should definitely be part of the PR.

> Then when we make a release, remember to also mention that and that support for the 2D Voronoi function has been dropped.

:+1:

kohr-h commented:
> I believe the NFFT utility functions (rand unit and voronoi) should be replaced by native ones using NumPy. 

Sounds reasonable, I'll make a new issue for it.

> Same for the solver, which could be substituted by wrapping a plan in a LinearOperator to be used within SciPy's implementation of the conjugate gradient.

Interesting. I haven't looked in detail, but if the NFFT solver really only does least squares, then any old optimizer from SciPy should do a better job. New issue coming up!

kohr-h commented:
I just ran all the unittests with a conda package of NFFT 3.5.0 (with single and long double), looking good. I could also test 3.4.x, but there won't be any conda packages with the required extra libs, so I wonder how useful such tests would be. Only for people who manually build NFFT.

grlee77 commented:
I think just 3.5 should be fine

grlee77 commented:
I tried this branch on OS X just now and get a segmentation fault on a test involving long double. Some long double tests for size (16, 16) passed prior to the failing one for size (8, 8). I will have to look into it further when I get a chance. I reran a couple times and it fails at the same test consistently.

I did:
1.) install nfft 3.5.0 from conda-forge
2.) checkout your branch and run `pip install -e . -v`
3.) `pytest pynufft/tests/ -v`

output:
```
======================================================== test session starts =========================================================
platform darwin -- Python 3.6.7, pytest-4.4.1, py-1.5.2, pluggy-0.9.0 -- /Users/lee8rx/anaconda/bin/python
cachedir: .pytest_cache
plugins: cov-2.6.1
collected 312 items                                                                                                                  

pynfft/tests/test_nfft.py::test_forward[ shapes=(8, 8) - use_dft=True - m=6 - prec='single' ] PASSED                           [  0%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(8, 8) - use_dft=True - m=6 - prec='single' ] PASSED                           [  0%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(8, 8) - use_dft=False - m=6 - prec='single' ] PASSED                          [  0%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(8, 8) - use_dft=False - m=6 - prec='single' ] PASSED                          [  1%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(8, 8) - use_dft=False - m=6 - prec='double' ] PASSED                          [  1%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(8, 8) - use_dft=False - m=6 - prec='double' ] PASSED                          [  1%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(16, 16) - use_dft=False - m=6 - prec='double' ] PASSED                        [  2%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(16, 16) - use_dft=False - m=6 - prec='double' ] PASSED                        [  2%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(16, 16) - use_dft=False - m=12 - prec='double' ] PASSED                       [  2%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(16, 16) - use_dft=False - m=12 - prec='double' ] PASSED                       [  3%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(16, 16) - use_dft=True - m=12 - prec='double' ] PASSED                        [  3%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(16, 16) - use_dft=True - m=12 - prec='double' ] PASSED                        [  3%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(16, 16) - use_dft=False - m=12 - prec='single' ] SKIPPED                      [  4%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(16, 16) - use_dft=False - m=12 - prec='single' ] SKIPPED                      [  4%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(16, 16) - use_dft=False - m=12 - prec='long double' ] PASSED                  [  4%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(16, 16) - use_dft=False - m=12 - prec='long double' ] PASSED                  [  5%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(16, 16) - use_dft=True - m=12 - prec='long double' ] PASSED                   [  5%]
pynfft/tests/test_nfft.py::test_adjoint[ shapes=(16, 16) - use_dft=True - m=12 - prec='long double' ] PASSED                   [  5%]
pynfft/tests/test_nfft.py::test_forward[ shapes=(8, 8) - use_dft=False - m=12 - prec='long double' ] Segmentation fault: 11
```

grlee77 commented:
Also, if I remove `long double` from the pytest fixture, then all remaining tests pass for single and double.

kohr-h commented:
That's interesting. I've seen the same failure occur on my Linux system when I added a `test` function that would run `pytest.main(['../', '-v'])` programatically. When run from the command line, everything is fine.

kohr-h commented:
> checkout your branch and run `pip install -e . -v`

How did it find `nfft.h` and the libraries?

grlee77 commented:
> How did it find nfft.h and the libraries?

I'm not sure, but it does:
```
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/lee8rx/anaconda/include -arch x86_64 -I/Users/lee8rx/anaconda/include -arch x86_64 -I/Users/lee8rx/anaconda/lib/python3.6/site-packages/numpy/core/include -I/Users/lee8rx/anaconda/include/python3.6m -c /Users/lee8rx/src/my_git/pyir/3rdparty/pynfft/pynfft/nfft.c -o build/temp.macosx-10.7-x86_64-3.6/Users/lee8rx/src/my_git/pyir/3rdparty/pynfft/pynfft/nfft.o -O3 -fomit-frame-pointer -fstrict-aliasing -ffast-math
```

I also tried again with the instructions from the readme and with GCC 5.4 vs. CLANG and got the same failure.

If I do a standard install instead of inplace one, calling `pytest --pyargs pynfft -v` gives the same result as well.

grlee77 commented:
Also, disregard my previous suggestion about using `api.h`. That file does not get installed as part of the package. It looks like the include folder (at least in the conda-forge package) only contains two files: `nfft3.h` and `nfft3mp.h`.

kohr-h commented:
> > How did it find nfft.h and the libraries?
> 
> I'm not sure, but it does:
> 
> ```
> clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/lee8rx/anaconda/include -arch x86_64 -I/Users/lee8rx/anaconda/include -arch x86_64 -I/Users/lee8rx/anaconda/lib/python3.6/site-packages/numpy/core/include -I/Users/lee8rx/anaconda/include/python3.6m -c /Users/lee8rx/src/my_git/pyir/3rdparty/pynfft/pynfft/nfft.c -o build/temp.macosx-10.7-x86_64-3.6/Users/lee8rx/src/my_git/pyir/3rdparty/pynfft/pynfft/nfft.o -O3 -fomit-frame-pointer -fstrict-aliasing -ffast-math
> ```
> 
> I also tried again with the instructions from the readme and with GCC 5.4 vs. CLANG and got the same failure.

Maybe that's just a difference between Linux and Mac OS in the way libraries and headers are discovered. Or how conda extends paths.
When I run the same commands, `nfft.h` is not found:

    building 'pynfft.nfft' extension
        gcc -pthread -B /home/hkohr/miniconda/envs/pynfft/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/hkohr/miniconda/envs/pynfft/lib/python3.7/site-packages/numpy/core/include -I/home/hkohr/miniconda/envs/pynfft/include/python3.7m -c /home/hkohr/git/pynfft/pynfft/nfft.c -o build/temp.linux-x86_64-3.7/home/hkohr/git/pynfft/pynfft/nfft.o -O3 -fomit-frame-pointer -malign-double -fstrict-aliasing -ffast-math
        /home/hkohr/git/pynfft/pynfft/nfft.c:624:10: fatal error: nfft3.h: No such file or directory
         #include "nfft3.h"
                  ^~~~~~~~~
        compilation terminated.

kohr-h commented:
> Also, disregard my previous suggestion about using `api.h`. That file does not get installed as part of the package. It looks like the include folder (at least in the conda-forge package) only contains two files: `nfft3.h` and `nfft3mp.h`.

Alright, makes sense.

kohr-h commented:
I tested building the package in-place and install with `pip install -e .`.
When I  run `pytest` from the repo root, everything is fine. When I exit the repo, though, and run the tests with `pytest --pyargs pynfft`, I also get a segfault.

To make testing a bit more independent from my machine, I added a `.travis.yml` and activated Travis CI on my fork. See [here](https://travis-ci.com/kohr-h/pyNFFT/builds/110621461) for the latest result. Maybe I can add OSX tests as well.
I think it would be a good addition in the main repo, too, but @ghisvail would need to activate it. That would also require some thinking about access tokens. Currently it would probably have to run via @ghisvail, but then we wouldn't be able to restart builds, for instance. Any idea here?

kohr-h commented:
### Update

See build: https://travis-ci.com/kohr-h/pyNFFT/builds/110624173

Python 2.7 builds and tests succeed. Not that I care a lot about Py2, but it's still interesting and odd. Makes me think about a warning Cython raises: 

    /home/travis/miniconda/envs/testenv/lib/python3.7/site-packages/Cython/Compiler/Main.py:367:
    FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2).
    This will change in a later release!
    File: /home/travis/build/kohr-h/pyNFFT/pynfft/nfft.pxd

Does that ring a bell? I searched the interweb a bit and found some hits, but I don't have a clear picture yet.

Pandas issue: https://github.com/pandas-dev/pandas/issues/23927
Pandas PR (check out the diff in `setup.py` and the import changes): https://github.com/pandas-dev/pandas/pull/24538

Cython documentation: [link](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html?highlight=language%20level) -- seems to imply that it mainly influences the string type, but apparently that's not the whole story.

grlee77 commented:
> Cython raises:
>
>/home/travis/miniconda/envs/testenv/lib/python3.7/site-packages/Cython/Compiler/Main.py:367:
>FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2).
>This will change in a later release!
>File: /home/travis/build/kohr-h/pyNFFT/pynfft/nfft.pxd
>Does that ring a bell? I searched the interweb a bit and found some hits, but I don't have a clear picture >yet.

Yes, I have dealt with that previously on another project. You can set it globally by passing `language_level=2`, to `cythonize` or set it at the top of each `.pyx` file. The easiest is to set it to `2` and the warning will go away . 

It is also probably not hard to update to `3`. I [switched to 3 for another project](https://github.com/PyWavelets/pywt/pull/435/files) and mainly had to change:
1.) the syntax of some import statements
2.) some usage of '/' to '//' to maintain integer division

grlee77 commented:
>  That would also require some thinking about access tokens. Currently it would probably have to run via @ghisvail, but then we wouldn't be able to restart builds, for instance. Any idea here?

I think it can be configured for the pyNFFT organization now that this repo was moved here from @ghisvail's personal account. He may need to set it up initially, but I think any organization members with appropriate permissions should be able to restart. I am not an expert on the CI/GitHub interaction, though, so I could be mistaken.

It would be very useful to have CI configured here.
